### PR TITLE
Reset global variables in python scripts when running entire script

### LINKED ISSUE #27185: Workbench and Mantidplot remembers the values of variables after executing a script
**Original Reporter:** James Lord

### Expected behavior ###
When I run a script it should start with a “clean” interpreter. Variables from the last run should not reappear unless they are set, and I should get a “name ‘x’ is not defined” error if I use a value before setting it, for example if I delete or comment out the line where it used to be set.

### Actual behavior ###
Variables defined in the script are recycled, causing different behaviour if the script is re-run after editing, compared to loading it later into a fresh tab and running.
Also modules loaded in an earlier run of the script persist if I comment out the “include” line. reload(module) doesn’t work very well if I’ve just edited the module in another tab, the traceback text-of-offending-line refers to the new version while my script actually called an earlier one.

### Steps to reproduce the behavior ###
Enter and run the following:

```
x=1
print (x)
```
and then comment out or delete the line “x=1” and run again.

What's intriguing is that MantidPlot does not show this behaviour despite having a 'Clear Variables' option which would suggest otherwise.

### Platforms affected ###
Mantid Workbench 4.1 on Windows, possibly others.

### PR DESCRIPTION
**Pull request by Harrietbrown**
**Description of work.**

This PR resets the global variable list in workbench python interpreter when running a whole script, this should not reset the variables if only running a partial script.

**Report to:** James Lord

**To test:**

1. Run the following python script
```
x=5
print(x)
```
2. remove the `x=5` line and highlight the print statment and run only that line. the x varable should still be kept.
3. Run the whole script without the `x=5` line, it should reset the varables so `x` will not be assigned.

Fixes #27185

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by NickDraper — CHANGES_REQUESTED** [review]
The changes are good, and work fine, but would it be possible to add a test for this behaviour into test_interpreter.py?

**Comment by Harrietbrown** [comment]
I've added a test for the functionality.

**Review by NickDraper — APPROVED** [review]
Great test added and works well, good to merge as soon as the builds are all happy.  Good work.

File: qt/python/mantidqt/widgets/codeeditor/test/test_interpreter.py [hunk 1]
```diff
@@ -56,6 +56,28 @@ def test_clear_key_binding(self):
                 self.assertEqual(w.clear_key_binding(key_combo), None,
                                  msg=fail_msg)
 
+    def test_variables_reset(self):
+        w = PythonFileInterpreter(content='x=\'this is a string\'\r\nprint(x)')
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/codeeditor/test/test_interpreter.py:60]: Nice test :+1:

**Pull request merged**